File: scripts_archive/node/root/tw_receiver.mjs
// Tiny HTTP server to receive scraped data from the browser
import { createServer } from 'http';
import { appendFileSync } from 'fs';

const OUTPUT_FILE = 'totalwine_lexington_green.jsonl';

const server = createServer((req, res) => {
  // CORS headers
  res.setHeader('Access-Control-Allow-Origin', '*');
  res.setHeader('Access-Control-Allow-Methods', 'POST, OPTIONS');
  res.setHeader('Access-Control-Allow-Headers', 'Content-Type');

  if (req.method === 'OPTIONS') {
    res.writeHead(204);
    res.end();
    return;
  }

  if (req.method === 'POST' && req.url === '/append') {
    let body = '';
    req.on('data', chunk => { body += chunk; });
    req.on('end', () => {
      appendFileSync(OUTPUT_FILE, body + '\n');
      const lines = body.split('\n').length;
      console.log(`Received ${lines} lines`);
      res.writeHead(200, { 'Content-Type': 'application/json' });
      res.end(JSON.stringify({ ok: true, lines }));
    });
    return;
  }

  if (req.url === '/done') {
    res.writeHead(200);
    res.end('done');
    console.log('All data received. Shutting down.');
    setTimeout(() => process.exit(0), 500);
    return;
  }

  res.writeHead(404);
  res.end('not found');
});

server.listen(9876, () => {
  console.log('Receiver listening on http://localhost:9876');
});

File: pipeline/analyze/tw_receiver.py
#!/usr/bin/env python3
"""
Tiny HTTP server to receive scraped data from the browser.
CORS-enabled, appends to JSONL file.

Usage:
    python -m pipeline.analyze.tw_receiver
"""

import sys
from http.server import HTTPServer, BaseHTTPRequestHandler
from pathlib import Path

sys.path.insert(0, str(Path(__file__).resolve().parents[2]))

OUTPUT_FILE = Path(__file__).resolve().parents[2] / "totalwine_lexington_green.jsonl"


class ReceiverHandler(BaseHTTPRequestHandler):
    def do_OPTIONS(self):
        self.send_response(204)
        self._cors_headers()
        self.end_headers()

    def do_POST(self):
        if self.path == "/append":
            length = int(self.headers.get("Content-Length", 0))
            body = self.rfile.read(length).decode("utf-8")
            with open(OUTPUT_FILE, "a", encoding="utf-8") as f:
                f.write(body + "\n")
            lines = body.count("\n") + 1
            print(f"Received {lines} lines")
            self.send_response(200)
            self._cors_headers()
            self.send_header("Content-Type", "application/json")
            self.end_headers()
            self.wfile.write(f'{{"ok": true, "lines": {lines}}}'.encode())
            return
        self.send_response(404)
        self.end_headers()
        self.wfile.write(b"not found")

    def do_GET(self):
        if self.path == "/done":
            self.send_response(200)
            self.end_headers()
            self.wfile.write(b"done")
            print("All data received. Shutting down.")
            import threading
            threading.Timer(0.5, lambda: sys.exit(0)).start()
            return
        self.send_response(404)
        self.end_headers()

    def _cors_headers(self):
        self.send_header("Access-Control-Allow-Origin", "*")
        self.send_header("Access-Control-Allow-Methods", "POST, OPTIONS")
        self.send_header("Access-Control-Allow-Headers", "Content-Type")


def main():
    server = HTTPServer(("", 9876), ReceiverHandler)
    print("Receiver listening on http://localhost:9876")
    server.serve_forever()


if __name__ == "__main__":
    main()
